models/event.py

```diff
@@ -145,34 +231,106 @@ class Meta:
     )
     ordering = ('date',)
 
+
 class PostbackURL(models.Model):
   event = models.ForeignKey('Event', on_delete=models.PROTECT, verbose_name='Event', null=False, blank=False, related_name='postbacks')
   url = models.URLField(blank=False,null=False,verbose_name='URL')
   class Meta:
     app_label = 'tracker'
 
+
 class SpeedRunManager(models.Manager):
   def get_by_natural_key(self, name, event):
     return self.get(name=name,event=Event.objects.get_by_natural_key(*event))
 
+
 class SpeedRun(models.Model):
   objects = SpeedRunManager()
   event = models.ForeignKey('Event', on_delete=models.PROTECT, default=LatestEvent)
   name = models.CharField(max_length=64,editable=False)
-  deprecated_runners = models.CharField(max_length=1024,blank=True,verbose_name='*DEPRECATED* Runners') # This field is now deprecated, we should eventually set up a way to migrate the old set-up to use the donor links
+  deprecated_runners = models.CharField(max_length=1024,blank=True,verbose_name='*DEPRECATED* Runners', editable=False) # This field is now deprecated, we should eventually set up a way to migrate the old set-up to use the donor links
   description = models.TextField(max_length=1024,blank=True)
-  starttime = models.DateTimeField(verbose_name='Start Time')
-  endtime = models.DateTimeField(verbose_name='End Time')
-  runners = models.ManyToManyField('Donor', blank=True, null=True)
+  starttime = models.DateTimeField(verbose_name='Start Time', editable=False, null=True)
+  endtime = models.DateTimeField(verbose_name='End Time', editable=False, null=True)
+  order = models.IntegerField(editable=False, null=True)  # can be temporarily null when moving runs around, or null when they haven't been slotted in yet
+  run_time = TimestampField(always_show_h=True)
+  setup_time = TimestampField(always_show_h=True)
+  runners = models.ManyToManyField('Runner')
   class Meta:
     app_label = 'tracker'
     verbose_name = 'Speed Run'
-    unique_together = ( 'name','event' )
-    ordering = [ 'event__date', 'starttime' ]
+    unique_together = (( 'name','event' ), ('event', 'order'))
+    ordering = [ 'event__date', 'order' ]
+
   def natural_key(self):
     return (self.name,self.event.natural_key())
+
   def clean(self):
-	if not self.name:
-	  raise ValidationError('Name cannot be blank')
+    if not self.name:
+      raise ValidationError('Name cannot be blank')
+
+  def save(self, fix_time=True, fix_runners=True, *args, **kwargs):
+    fix_time = self.order and self.run_time and self.setup_time and fix_time # we can't fix the time without all of these
```

Comment:
I just meant that if I want to create a test data, to avoid having to fill setup_time. In retrospect, its rather petty and insignificant, so don't worry about it.